File: ml-service/schemas/prediction.py
from pydantic import BaseModel, ConfigDict, Field


class PredictionRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    temperature: float = Field(ge=-10, le=50)
    oxygenLevel: float = Field(ge=10, le=30)
    humidity: float = Field(ge=0, le=100)
    windSpeed: float = Field(ge=0, le=50)
    pressure: float = Field(ge=900, le=1050)
    rainfall: float = Field(ge=0, le=500)


class PredictionResponse(BaseModel):
    prediction: int
    probability: float
    probabilities: dict[str, float]
    model_version: str

File: backend/src/models/Prediction.js
import mongoose from "mongoose";

const predictionSchema = new mongoose.Schema(
  {
    user: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true, index: true },
    inputs: {
      temperature: { type: Number, required: true },
      oxygenLevel: { type: Number, required: true },
      humidity: { type: Number, required: true },
      windSpeed: { type: Number, required: true },
      pressure: { type: Number, required: true },
      rainfall: { type: Number, required: true },
    },
    prediction: { type: Number, required: true, enum: [0, 1] },
    probability: { type: Number, required: true, min: 0, max: 1 },
    probabilities: { type: Map, of: Number, required: true },
    modelVersion: { type: String, required: true },
  },
  { timestamps: true },
);

export const Prediction = mongoose.model("Prediction", predictionSchema);
